### ISSUE TITLE
Remove outlier images extracted in Colab ....because they have no dots to refine

### ISSUE THREAD

**Issue by kaylago** [DeepLabCut/DeepLabCut#1848]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

I extracted outlier images from analyzed videos in Colab before realizing that doing so in Colab would result in errors - the images extracted in Colab have no labels on them to refine. I thought I could extract the images in Colab (as it would be faster with a GPU) and then refine the labels on my CPU, but the images didn't extract correctly. I would like to simply delete the outlier frames and their corresponding rows from the machine-labels h5 and machine-labels csv files, however I’m not sure what function to use to do this. Simply deleting the extra images and then running the function deeplabcut.dropannotationfileentriesduetodeletedimages(config) does not work in this case…Is there an alternate function I could use? Thanks!!

### Operating System

Windows/Google Colab

### DeepLabCut version

dlc version 2.2.1 on colab
dlc version 2.2.0.6 on windows

### DeepLabCut mode

single animal

### Device type

gpu google for outlier image extraction
cpu on windows when I ran the refine labels function

### Steps To Reproduce

1. Windows code:
(base) C:\Users\kkmgo>activate DEEPLABCUT

(DEEPLABCUT) C:\Users\kkmgo>ipython
Python 3.8.12 | packaged by conda-forge | (default, Oct 12 2021, 21:22:46) [MSC v.1916 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.1.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import deeplabcut
2022-05-25 08:31:29.363636: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'cudart64_110.dll'; dlerror: cudart64_110.dll not found
2022-05-25 08:31:29.364225: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.

In [2]: deeplabcut.__version__
Out[2]: '2.2.0.6'

In [3]: config_path='C:/Users/kkmgo/Dropbox/Turtle_Dance-Kayla-2022-01-26/config.yaml'

In [4]: deepla
![Screenshot 2022-05-25 102924](https://user-images.githubusercontent.com/36450185/170286740-ffe51c11-d5d2-4f42-9e5f-eb0ab9c7cc6a.png)
bcut.refine_labels(config_path) #attached photo of the images brought up in the GUI

3. config is pasted below:
  # Project definitions (do not edit)
Task: Turtle_Dance
scorer: Kayla
date: Jan26
multianimalproject: false
identity:

    # Project path (change when moving around)
project_path: /content/drive/My Drive/Turtle_Dance-Kayla-2022-01-26

    # Annotation data set configuration (and individual video cropping parameters)
video_sets:
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L201_DEC08.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L203_DEC08.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L205_DEC08.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L206_DEC08.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L201_DEC16.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L203_DEC16.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L212_DEC16.mp4:
    crop: 0, 1440, 0, 1920
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L214_DEC16.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L216_DEC16.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L202_DEC17.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L204_DEC17.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L207_DEC17.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L209_DEC17.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L211_DEC17.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L213_DEC17.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L201_DEC18.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L203_DEC18.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L205_DEC18.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L206_DEC18.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L208_DEC18.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L212_DEC18.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L214_DEC18.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\Turtle_Dance-Kayla-2022-01-26\videos\L216_DEC18.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\DLC-VideostoAnalyze\L203_DEC16.mp4:
    crop: 0, 1920, 0, 1440
  C:\Users\kkmgo\Dropbox\DLC-VideostoAnalyze\L214_DEC16.mp4:
    crop: 0, 1920, 0, 1440
  /content/drive/My Drive/DLC-VideostoAnalyze/Analyzed/L203_DEC16.mp4: #want to remove
    crop: 0, 1920, 0, 1440
  /content/drive/My Drive/DLC-VideostoAnalyze/Analyzed/L212_DEC16.mp4:  #want to remove
    crop: 0, 1440, 0, 1920
  /content/drive/My Drive/DLC-VideostoAnalyze/Analyzed/L205_DEC16.mp4:  #want to remove
    crop: 0, 1920, 0, 1440
bodyparts:
- headanterior
- headposterior
- neck
- nuchalscute
- vertscute1
- vertscute2
- vertscute3
- vertscute4
- vertscute5
- carapaceposterior
- Lshoulder
- LFFproximal
- LFFmid
- LFFdistal
- Rshoulder
- RFFproximal
- RFFmid
- RFFdistal
- LRF
- RRF
- nares
- Lsupraorbital
- Rsupraorbital
- Leye
- Reye
- mouthopen

    # Fraction of video to start/stop when extracting frames for labeling/refinement

    # Fraction of video to start/stop when extracting frames for labeling/refinement
start: 0
stop: 1
numframes2pick: 10

    # Plotting configuration
skeleton:
- - headanterior
  - headposterior
- - neck
  - nuchalscute
- - vertscute1
  - vertscute2
- - vertscute3
  - vertscute4
- - vertscute5
  - carapaceposterior
- - Lshoulder
  - LFFproximal
- - LFFmid
  - LFFdistal
- - Rshoulder
  - RFFproximal
- - RFFmid
  - RFFdistal
- - LRF
  - RRF
- - nares
  - Lsupraorbital
- - Rsupraorbital
  - Leye
- - Reye
  - mouthopen

skeleton_color: black
pcutoff: 0.8
dotsize: 4
alphavalue: 0.7
colormap: rainbow

    # Training,Evaluation and Analysis configuration
TrainingFraction:
- 0.95
iteration: 1
default_net_type: resnet_50
default_augmenter: default
snapshotindex: -1
batch_size: 1

    # Cropping Parameters (for analysis and outlier frame detection)
cropping: false
    #if cropping is true for analysis, then set the values here:
x1: 0
x2: 640
y1: 277
y2: 624

    # Refinement configuration (parameters from annotation dataset configuration also relevant in this stage)
corner2move2:
- 50
- 50
move2corner: true
![image](https://user-images.githubusercontent.com/36450185/170286835-5d4c7a53-7913-49af-9c74-52d44bfec615.png)

### Relevant log output

```shell
NA - everything is above
```

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by jeylau** [comment]
Hi @kaylago! I've just tried the use case you described, and had no problem extracting outlier frames in Colab and label them locally. Is your machinelabels.csv file empty? If not, could it be that point size is so small that hollow points cannot be seen on the image? You could try with a lower pcutoff threshold (e.g., 0.1) so dots appear filled.
Anyway, you could manually delete the machinelabels h5 and csv files, and call `deeplabcut.dropimagesduetolackofannotation` to drop the images that were not originally annotated: this should leave you in the state just prior to outlier extraction.
(Also note that the presence of a GPU is unlikely to accelerate outlier extraction: it is almost the same procedure as frame extraction, and is mostly I/O operations on individual image files.)

**Comment by kaylago** [comment]
Yes the machine labels.csv file is empty for those images. I had extracted one round of outlier images (using my CPU), fixed the labels, retrained, and then extracted a second round of outlier images (using COLAB), and the second round is what caused the issue. So perhaps it is because I tried to extract from the same video twice?
[machinelabels.csv](https://github.com/DeepLabCut/DeepLabCut/files/8798938/machinelabels.csv)
!

**Comment by jeylau** [comment]
Extracting multiple times from the same video should work, I've just tried it 5 times 😄. No rows are empty in the machinelabels file though, so I don't understand what happened in your case. Did you extract outlier frames automatically or manually with the GUI? And I'm curious: you were still not happy with your model performance after the first round of retraining?

**Comment by kaylago** [comment]
I extracted the outlier frames automatically in Colab and that's where I got blank rows! I was not happy with some of the machine label placements - the turtles are at an air-water interface so, I think that's making it tricky. Also apologies for the delayed reply - got caught up with field work! Thanks for your help so far!

**Comment by kaylago** [comment]
The function deeplabcut.dropimagesduetolackofannotation isn't working, nor is deeplabcut.dropannotationfileentriesduetodeletedimages on the machinelabels csv and h5 files. I think that perhaps the frames extracted simply do not have labels, because I've noticed that there are not constant labels appearing in the videos - there is a lot of blinking of the dots in the labeled video. Is there another solution to get rid of these frames and their blank rows in the machinelabels files?

![image](https://user-images.githubusercontent.com/36450185/189725637-080bb59a-bc42-4eed-8e98-e60a314e06fc.png)

**Comment by jeylau** [comment]
Hi @kaylago, if you update your Windows installation of DLC to at least 2.2.1, the function won't raise the AttributeError, it was fixed by a user with #1751.
